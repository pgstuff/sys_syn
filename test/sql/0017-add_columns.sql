BEGIN;

CREATE EXTENSION tinyint SCHEMA public;
CREATE EXTENSION sys_syn;

CREATE SCHEMA user_data
    AUTHORIZATION postgres;

CREATE TABLE user_data.test_table (
        test_table_id integer NOT NULL,
        test_table_text text,
        CONSTRAINT test_table_pid PRIMARY KEY (test_table_id));

INSERT INTO sys_syn.in_groups_def VALUES ('in');

SELECT sys_syn.in_table_create_sql('user_data.test_table'::regclass, 'in');

DO $$BEGIN
    EXECUTE sys_syn.in_table_create_sql('user_data.test_table'::regclass, 'in');
END$$;

INSERT INTO user_data.test_table(
        test_table_id, test_table_text)
VALUES (1,              'test_data 1');

INSERT INTO sys_syn.out_groups_def VALUES ('out');

SELECT sys_syn.out_table_create('user_data', 'test_table', 'out', data_view => TRUE);

SELECT user_data.test_table_pull(FALSE);
SELECT user_data.test_table_out_move_1();

SELECT * FROM user_data.test_table_out_queue_data_1;

ALTER TABLE user_data.test_table
        ADD COLUMN test_table_date date,
        ADD COLUMN test_table_datetime timestamp with time zone;

SELECT sys_syn.in_table_columns_add_sql('test_table');

DO $$BEGIN
    EXECUTE sys_syn.in_table_columns_add_sql('test_table', 'user_data.test_table'::regclass);
END$$;

SELECT * FROM user_data.test_table_out_queue_data_1;

INSERT INTO user_data.test_table(
        test_table_id, test_table_text,        test_table_date,        test_table_datetime)
VALUES (2,              'test_data 2',          '6283-01-08',           '6283-01-08 05:30:07');

UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;SET LOCAL sys_syn.trans_id_curr TO 2;
SELECT user_data.test_table_pull(FALSE);
SELECT user_data.test_table_out_move_1();

SELECT * FROM user_data.test_table_out_queue_data_1;

ROLLBACK;
