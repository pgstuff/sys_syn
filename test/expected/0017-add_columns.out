BEGIN;
CREATE EXTENSION tinyint
    SCHEMA public;
CREATE EXTENSION sys_syn;
CREATE SCHEMA user_data
    AUTHORIZATION postgres;
CREATE TABLE user_data.test_table (
        test_table_key integer NOT NULL,
        test_table_text text,
        CONSTRAINT test_table_pkey PRIMARY KEY (test_table_key));
INSERT INTO sys_syn.in_groups_def VALUES ('in');
SELECT sys_syn.in_table_add_sql('user_data.test_table'::regclass, 'in');
                                           in_table_add_sql                                            
-------------------------------------------------------------------------------------------------------
 SELECT sys_syn.in_table_add (                                                                        +
                 schema          => 'user_data',                                                      +
                 in_table_id     => 'test_table',                                                     +
                 in_group_id     => 'in',                                                             +
                 in_pull_id      => NULL,                                                             +
                 in_columns      => ARRAY[                                                            +
                        $COL$("test_table_key","integer",Key,"in_source.test_table_key",,,,)$COL$,    +
                        $COL$("test_table_text","text",Attribute,"in_source.test_table_text",,,,)$COL$+
                 ]::sys_syn.create_in_column[],                                                       +
                 full_table_reference    => 'user_data.test_table',                                   +
                 changes_table_reference => NULL,                                                     +
                 full_sql                => NULL,                                                     +
                 changes_sql             => NULL,                                                     +
                 full_pre_sql            => NULL,                                                     +
                 changes_pre_sql         => NULL,                                                     +
                 full_post_sql           => NULL,                                                     +
                 changes_post_sql        => NULL,                                                     +
                 enable_deletes_implied  => TRUE,                                                     +
                 full_prepull_id         => NULL,                                                     +
                 changes_prepull_id      => NULL                                                      +
         );
(1 row)

DO $$BEGIN
    EXECUTE sys_syn.in_table_add_sql('user_data.test_table'::regclass, 'in');
END$$;
INSERT INTO user_data.test_table(
        test_table_key, test_table_text)
VALUES (1,              'test_data 1');
INSERT INTO sys_syn.out_groups_def VALUES ('out');
SELECT sys_syn.out_table_add('user_data', 'test_table', 'out', data_view => TRUE);
 out_table_add 
---------------
 
(1 row)

SELECT user_data.test_table_pull(FALSE);
 test_table_pull 
-----------------
 t
(1 row)

SELECT user_data.test_table_out_move();
 test_table_out_move 
---------------------
 t
(1 row)

SELECT * FROM user_data.test_table_out_queue_data;
 sys_syn_key | sys_syn_trans_id_in | sys_syn_delta_type | sys_syn_queue_state | sys_syn_queue_id | sys_syn_queue_priority | sys_syn_hold_updated | sys_syn_hold_trans_id_first | sys_syn_hold_trans_id_last | sys_syn_hold_reason_count | sys_syn_hold_reason_id | sys_syn_hold_reason_text | sys_syn_trans_id_out | sys_syn_processed_time | test_table_key | test_table_text 
-------------+---------------------+--------------------+---------------------+------------------+------------------------+----------------------+-----------------------------+----------------------------+---------------------------+------------------------+--------------------------+----------------------+------------------------+----------------+-----------------
 (1)         |                   2 | Add                | Unread              |                  |                        |                      |                             |                            |                           |                        |                          |                      |                        |              1 | test_data 1
(1 row)

ALTER TABLE user_data.test_table
        ADD COLUMN test_table_date date,
        ADD COLUMN test_table_datetime timestamp with time zone;
SELECT sys_syn.in_table_columns_add_sql('test_table');
                                                     in_table_columns_add_sql                                                      
-----------------------------------------------------------------------------------------------------------------------------------
 SELECT sys_syn.in_table_columns_add (                                                                                            +
                 'test_table',                                                                                                    +
                 ARRAY[                                                                                                           +
                        $COL$("test_table_date","date",Attribute,"in_source.test_table_date",,,,)$COL$,                           +
                        $COL$("test_table_datetime","timestamp with time zone",Attribute,"in_source.test_table_datetime",,,,)$COL$+
                 ]::sys_syn.create_in_column[]                                                                                    +
         );
(1 row)

DO $$BEGIN
    EXECUTE sys_syn.in_table_columns_add_sql('test_table', 'user_data.test_table'::regclass);
END$$;
SELECT * FROM user_data.test_table_out_queue_data;
 sys_syn_key | sys_syn_trans_id_in | sys_syn_delta_type | sys_syn_queue_state | sys_syn_queue_id | sys_syn_queue_priority | sys_syn_hold_updated | sys_syn_hold_trans_id_first | sys_syn_hold_trans_id_last | sys_syn_hold_reason_count | sys_syn_hold_reason_id | sys_syn_hold_reason_text | sys_syn_trans_id_out | sys_syn_processed_time | test_table_key | test_table_text | test_table_date | test_table_datetime 
-------------+---------------------+--------------------+---------------------+------------------+------------------------+----------------------+-----------------------------+----------------------------+---------------------------+------------------------+--------------------------+----------------------+------------------------+----------------+-----------------+-----------------+---------------------
 (1)         |                   2 | Add                | Unread              |                  |                        |                      |                             |                            |                           |                        |                          |                      |                        |              1 | test_data 1     |                 | 
(1 row)

INSERT INTO user_data.test_table(
        test_table_key, test_table_text,        test_table_date,        test_table_datetime)
VALUES (2,              'test_data 2',          '6283-01-08',           '6283-01-08 05:30:07');
UPDATE sys_syn.trans_id_mod SET trans_id_mod = trans_id_mod + 1;
SELECT user_data.test_table_pull(FALSE);
 test_table_pull 
-----------------
 t
(1 row)

SELECT user_data.test_table_out_move();
 test_table_out_move 
---------------------
 t
(1 row)

SELECT * FROM user_data.test_table_out_queue_data;
 sys_syn_key | sys_syn_trans_id_in | sys_syn_delta_type | sys_syn_queue_state | sys_syn_queue_id | sys_syn_queue_priority | sys_syn_hold_updated | sys_syn_hold_trans_id_first | sys_syn_hold_trans_id_last | sys_syn_hold_reason_count | sys_syn_hold_reason_id | sys_syn_hold_reason_text | sys_syn_trans_id_out | sys_syn_processed_time | test_table_key | test_table_text | test_table_date |     test_table_datetime      
-------------+---------------------+--------------------+---------------------+------------------+------------------------+----------------------+-----------------------------+----------------------------+---------------------------+------------------------+--------------------------+----------------------+------------------------+----------------+-----------------+-----------------+------------------------------
 (1)         |                   3 | Add                | Unread              |                  |                        |                      |                             |                            |                           |                        |                          |                      |                        |              1 | test_data 1     |                 | 
 (2)         |                   3 | Add                | Unread              |                  |                        |                      |                             |                            |                           |                        |                          |                      |                        |              2 | test_data 2     | 01-08-6283      | Mon Jan 08 05:30:07 6283 PST
(2 rows)

ROLLBACK;
